### ISSUE TITLE
RFC: Add linear type facility

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rfcs#814]
[Rendered](https://github.com/aidancully/rfcs/blob/linear-trait/text/0000-linear-type.md)

Tracking issue for postponed PR #776

**Comment by mzabaluev** [comment]
This may be useful to support `close` methods for I/O objects that need error reporting when being disposed of. See [a discussion thread](http://internals.rust-lang.org/t/issues-in-new-i-o/1658) on I/O issues.

**Comment by Munksgaard** [comment]
Are there any plans for when we take this up again? 

#523 seems to be related, by the way.

**Comment by aidancully** [comment]
@Munksgaard I've factored some of this RFC out into #1180, and (opened a discussion on internals about another part)[https://internals.rust-lang.org/t/pre-rfc-excdrop-trait-for-different-drop-glue-during-unwinding/2242]. I've been working through an implementation of #1180, but real life is taking too much time to make any progress at all for probably the next month or so.

**Comment by Rufflewind** [comment]
Aside from fallible `Drop` constructors (e.g. closing a file), there’s another situation where it could come in handy:  Sometimes an resource is held by another thread/process/host and dropping would require waiting for for the other side to complete whatever they are doing.  Implicit drops would cause cause the program to block, possibly unintentionally (which can further lead to unintended deadlocks).  Worse, the protocol might not even offer a way to wait for it, in which case there’s nothing you can do but panic.  In these kinds of cases, it would have been better to have simply caught it as a compile error.

I think linear types are suited for transient, active things that need to be dealt with carefully and explicitly, whereas affine types are more like passive objects that could be thrown away without regard.  Which one is more suitable for a particular resource is probably a bit subjective and dependent on the robustness and reliability demands.

It might even be possible to have a rudimentary lint-based implementation of linear types: if an object gets dropped and the destructor unconditionally panics, then that’s almost certainly a bug and should be warned.

**Comment by jethrogb** [comment]
> It might even be possible to have a rudimentary lint-based implementation of linear types: if an object gets dropped and the destructor unconditionally panics, then that’s almost certainly a bug and should be warned.

How would that work? Presumably when people say linear types they don't mean types with values with an infinite lifetime, but rather that you need to pass them to some special destroy function by value when you're done with them. If you have a `Drop` impl that always panics you can't even do this.

**Comment by Rufflewind** [comment]
> How would that work? Presumably when people say linear types they don't mean types with values with an infinite lifetime, but rather that you need to pass them to some special destroy function by value when you're done with them.

It’s not infinitely long, but the time at which it becomes destructible is not statically known, and may require the user to take certain actions before it becomes destructible.

> If you have a `Drop` impl that always panics you can't even do this (safely).

Yes, [panicking in `drop` is really awkward](https://github.com/bsteinb/rsmpi/pull/18#issuecomment-289942053), so I’m hoping linear types could help obviate the need for such workarounds.

**Comment by LiamTheProgrammer** [comment]
It's been over 5 years now since this was postponed.

**Comment by carado** [comment]
6 years now, is this ever gonna be considered ?
There have been a bunch of times when I've wanted something like this.

**Comment by Philonoist** [comment]
This is so important! If this would be supported, rust could be the primary language to compile to cairo-lang...

**Comment by leviska** [comment]
Use cases:
1. Async drop. Instead of implementing async drop, you can forbid drop for the type and implement async fn method that must be called instead of dropping.
2. Drop with arguments. Imagine a handle, that is created from some resource manager and needs a reference to this manager on drop. One way is to store the reference inside the handle (which can be tricky/expensive with Arcs/Mutexes/etc.), but the other way is to make drop accept the reference to the resource manager. With this you can forbid drop and implement custom function that accepts additional arguments.
3. Drop that can fail/can have side effects. Again, the same idea, you can forbid drop and write a function, which will return error, for example, if drop failed, so it will be the job of the programmer to handle this error correctly

**Comment by Demindiro** [comment]
I've found that using `must_use` in combination with `forbid(unused_must_use)` is effective as a crappy form of linear types, e.g.:

```rust
#![forbid(unused_must_use)]

#[derive(Debug)]
#[must_use = "Must be manually destroyed with `T::destroy()`"]
struct T;

impl T {
    fn destroy(self) {
        std::mem::forget(self)
    }
}

impl Drop for T {
    fn drop(&mut self) {
        if !std::thread::panicking() {
            panic!("Must be destroyed with `T::destroy()`");
        }
    }
}

fn main() {
    dbg!(T);
}
```

```
error: unused `T` that must be used
  --> src/main.rs:22:5
   |
22 |     dbg!(T);
   |     ^^^^^^^
   |
   = note: Must be manually destroyed with `T::destroy()`
```

If you replace `dbg!(T)` with `dbg!(T).destroy()` it compiles and works as expected.

**Comment by SOF3** [comment]
Where does the word "linear" come from? Is there any prior work in other languages for reference?

**Comment by Kixiron** [comment]
Linear types are a form of [substructural type systems](https://en.wikipedia.org/wiki/Substructural_type_system#Linear_type_systems), notable implementors are Clean and Haskell (via GHC extension)

**Comment by afetisov** [comment]
@SOF3  The word "linear" comes from the notion of [linear logic](https://en.wikipedia.org/wiki/Linear_logic). I believe the concept, and the terminology, were first proposed in a 1987 paper _Linear Logic_  by J.-Y. Girard. The terminology was likely motivated by two related concepts, coming from the theory of vector spaces and linear operators. People often talk about "linear relations" or "linear terms" when variables occur only once in the formula, by analogy with [linear functions](https://en.wikipedia.org/wiki/Linear_function), which have a single occurrence of each coordinate variable in their expression.  In linear logic, an element of each type can be used only once in the formula, because construction of pairs `(x1, x2): X \times X` is forbidden for general types. Linear logic doesn't allow duplication of values (unless special operators are used). Similarly, linear logic doesn't allow silent discarding of values: if you have an element `x: X`, you must explicitly thread it through all formulas, and can't just "forget" to use it.

The second, more conceptual, reason to call this logic "linear" is that it arises as [internal logic](https://ncatlab.org/nlab/show/internal+logic) of the [additive](https://en.wikipedia.org/wiki/Additive_category) [symmetric monoidal](https://en.wikipedia.org/wiki/Symmetric_monoidal_category) [category of vector spaces](https://en.wikipedia.org/wiki/Category_of_modules). Basically it is the only way your logic could be structured if vector spaces were all you knew about, and it allows to prove all facts about them. The categorical model served as the guiding star for the formulation of logic, but for practical applications the later found model of [resource semantics](https://en.wikipedia.org/wiki/Linear_logic#The_resource_interpretation) is much more relevant.

**Comment by aleokdev** [comment]
Any progress on this? What is it blocked on? This would be incredibly useful for interfacing with externally handled buffers, such as the ones present in a GPU. Other alternatives require having a device around to be able to destroy the resources, which unnecessarily adds bloat to the structure or to the global namespace.

I'd be willing to work on it if the PR gets considered.

**Comment by golddranks** [comment]
@aleokdev Unlikely to happen. I think @Gankra's blog post "The Pain Of Real Linear Types in Rust", that for some reason is not linked in this thread, killed the feature for many: https://faultlore.com/blah/linear-rust/

**Comment by psionic-k** [comment]
I will argue that we are overemphasizing the cost of an overly ambitious implementation when the real implementation we need supports almost nothing but prevents the very real problems of in-flight things being silently dropped.  The contract we aim to create is not an unreasonable one, that no envelope may be licked unless it is sealed and sent.

> https://faultlore.com/blah/linear-rust/
> There would need to be some special handling if <T: Copy>

It may not be a contradiction to split the licked envelope so long as both parts are eventually sealed and sent.   However, if the semantics of `MustConsume` tend to involve some external service holding an imperiled `LickedEnvelope`, isn't it more natural to expect the user to tell the external service to split the real external envelope before requiring the program to express that both parts of its representation will be consumed?  If so, isn't it natural for the compiler to require `Clone` and `Copy` to be manually implemented for any `MustConsume` since the user will need to notify the external service before any two such imperiled `LickedEnvelopes` can be made to exist?  Wouldn't the compiler's refusal to implement `MustConsume` and `Copy` automatically be a strong suggestion to the user that these two ideas are orthogonal, and isn't that suggestion accurate?

> A bunch of the standard library will need to be “upgraded” to have ?Leave support. 

I don't buy this.  I would be fine with all of std disallowing `MustConsume` the same way many types disallow all but `Sized`.   It is easy enough to hack away `MustConsume` temporarily with some other outer `MustConsume` contract on a container that uses `Vec<MustConsumeErased>` internally.  If I'm writing APIs for the user, this maneuver is merely some harmless boilerplate that a derive crate would swiftly conceal while maintaining soundness.

I think there's a real case that we're tunnel-visioning on the hard implications that are not meaningful to actual use cases for `MustConsume` while ignoring the real harm to users when imperiled `LickedEnvelopes` are silently dropped or only explode at runtime in the rare paths that were meant to recover.  Like thread locals or branded lifetimes, `MustConsume` is the kind of *semantic* that I want automation for when the flow control is naturally going to constrain the problem to a much smaller surface to defend.

> People who say they want Linear™ Types In Rust actually just want Proper Support For Relevant™ Types.

In the case of a panic, I may still want to implement custom `Drop` behavior that may clean up a runtime condition my program was not allowed to express, enabling imperiled `LickedEnvelope` to be rescued on the remote service.  If this counts as using my type at-least-once (once for the `MustConsume` that may have failed and once in a panic hook) then I would consider Relevant types sufficient.

> This would be incredibly useful for interfacing with externally handled buffers, such as the ones present in a GPU.

I have encountered this need when dealing with this same need for a contract.  In stable Rust, I can warn and I can explode for the user at runtime, and neither one of these things are correct.  When we swear to do it all right in the name of correctness, let us not forget the errors that already happen and the robustness of conservative solutions that only support the vary narrow case.

What I'm looking at in the meantime is [this perhaps dirty thing](https://internals.rust-lang.org/t/an-approach-to-linear-ish-types/21111/4).  If it saves my users from writing programs that drop an already recorded command buffer and no other implementation would demand any different semantics on the public API, that's a win.

**Comment by Hisanatos** [comment]
Out of pure curiosity, because I'm just a normal programmer usually using some programming language, and not fully understanding why it works or how.
Wasn't this accepted here https://github.com/rust-lang/rust-project-goals/blob/main/src/2026/move-trait.md as a rust goal and isn't it being tracked here? https://github.com/rust-lang/rust-project-goals/issues/635

As far as I understand linear types differs from Rust implementation in this way
Rust: At most once
Linear: Exactly once.

Wouldn't auto !Forget trait solve this problem?
